src/NHibernate.Test/NHSpecificTest/GH2029/Fixture.cs

```diff
@@ -0,0 +1,181 @@
+﻿using System;
+using System.Linq;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Dialect;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH2029
+{
+	public class TestClass
+	{
+		public virtual int Id { get; set; }
+		public virtual int? NullableInt32Prop { get; set; }
+		public virtual int Int32Prop { get; set; }
+		public virtual long? NullableInt64Prop { get; set; }
+		public virtual long Int64Prop { get; set; }
+	}
+
+	[TestFixture]
+	public class Fixture : TestCaseMappingByCode
+	{
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			mapper.Class<TestClass>(rc =>
+			{
+				rc.Id(x => x.Id, m => m.Generator(Generators.Native));
+				rc.Property(x => x.NullableInt32Prop);
+				rc.Property(x => x.Int32Prop);
+				rc.Property(x => x.NullableInt64Prop);
+				rc.Property(x => x.Int64Prop);
+			});
+
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+		protected override bool AppliesTo(Dialect.Dialect dialect)
+		{
+			return !(dialect is SQLiteDialect);
+		}
+
+		protected override void OnSetUp()
+		{
+			using (var session = OpenSession())
+			using (var tx = session.BeginTransaction())
+			{
+				session.Save(new TestClass
+				{
+					Int32Prop = int.MaxValue,
+					NullableInt32Prop = int.MaxValue,
+					Int64Prop = int.MaxValue,
+					NullableInt64Prop = int.MaxValue
+				});
+				session.Save(new TestClass
+				{
+					Int32Prop = int.MaxValue,
+					NullableInt32Prop = int.MaxValue,
+					Int64Prop = int.MaxValue,
+					NullableInt64Prop = int.MaxValue
+				});
+				session.Save(new TestClass
+				{
+					Int32Prop = int.MaxValue,
+					NullableInt32Prop = null,
+					Int64Prop = int.MaxValue,
+					NullableInt64Prop = null
+				});
+
+				tx.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var tx = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from TestClass").ExecuteUpdate();
+
+				tx.Commit();
+			}
+		}
+
+		[Test]
+		public void NullableIntOverflow()
+		{
+			using (var session = OpenSession())
+			using (session.BeginTransaction())
+			using (var sqlLog = new SqlLogSpy())
+			{
+				var groups = session.Query<TestClass>()
+					   .GroupBy(i => 1)
+					   .Select(g => new
+					   {
+						   s = g.Sum(i => (long) i.NullableInt32Prop)
+					   })
+					   .ToArray();
```

Comment:
Has `ToArray` be used on purpose to avoid async generation?
Since we may someday support `ToArrayAsync`, I think we should use `ToList` here for now, in order to avoid changes here if `ToArrayAsync` is added.